raytracing/matrixgroup.py

```diff
@@ -235,3 +235,119 @@ def __next__(self):
             self.iteration += 1
             return element
         raise StopIteration
+
+    def __len__(self):
+        return len(self.elements)
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):  # If we get a slice, return a matrix group
+            return MatrixGroup(self.elements[item])
+        return self.elements[item]
+
+    def removeElement(self, index: int, pad: bool = False):
+        maxIndex = len(self)
+        if index < 0:
+            index += maxIndex
+        if index >= maxIndex or index < 0:
+            raise IndexError(f"Index {index} out of bound, min = 0, max {maxIndex - 1}.")
+        tempElements = self.elements[:]
+        self.elements = []
+        for i in range(maxIndex):
+            element = tempElements[i]
+            if i == index:
+                length = element.L
+                if length > 0 and pad:  # No need to pad if length is null
+                    element = Space(length)
+                    self.append(element)
+            else:
+                self.append(element)
+
+    def insertElement(self, index: int, element: Matrix):
+        if not isinstance(element, collections.Iterable):
+            element = MatrixGroup([element])
+        else:
+            element = MatrixGroup(element)
+        maxIndex = len(self)
+        if index < 0:
+            index += maxIndex
+        if index > maxIndex or index < 0:
+            raise IndexError(f"Index {index} out of bound, min = 0, max {maxIndex}.")
+        if index == maxIndex:
+            for newElement in element:
+                self.append(newElement)
+        else:
+            tempElements = self.elements[:]
+            self.elements = []
+            for i in range(maxIndex):
+                previousElement = tempElements[i]
+                if i == index:
+                    for newElement in element:
+                        self.append(newElement)
+                    self.append(previousElement)
+                else:
+                    self.append(previousElement)
+
+    def replaceSingle(self, index: int, newElement: Matrix):
```

Comment:
Can't this be done with `__getitem__` ?

Reply:
No because we need `__setitem__` to support assignment. That's why I thought of ways to support it. We can't just do
```python
mg = MatrixGroup([Space(10), Lens(5), Space(10)])
mg[1] = Lens(10)
```
with only `__getitem__`.